File: go/format.go
package main

// The reply contract, shared by every bot in this repository: the TypeScript
// and Python versions print byte-identical strings from the same fixtures, and
// the tests in each language assert it. Change one, change all three.

import (
	"fmt"
	"slices"
	"strings"

	"github.com/pokemontcgapi/sdk-go"
)

// One Cardmarket row per print language: an English copy and a French copy of
// the same card are two listings, not one price. The trial plan serves the
// English row only and says so in meta.withheld.
var locales = []string{"en", "de", "fr", "es", "it", "pt", "ja", "zh", "ko"}

const noQuote = "no quote today"

func title(card *pokemontcgapi.Card) string {
	return fmt.Sprintf("%s · %s #%s", card.Name, card.SetName, card.Number)
}

// raw keeps the ungraded rows only. A graded row is a different market, and on
// the API it is a different plan: a bigger bot gives slabs their own command.
func raw(quotes []pokemontcgapi.Price) []pokemontcgapi.Price {
	var out []pokemontcgapi.Price
	for _, q := range quotes {
		if q.Grading == nil {
			out = append(out, q)
		}
	}
	return out
}

func locale(q pokemontcgapi.Price) string {
	if q.Locale == nil {
		return ""
	}
	return *q.Locale
}

func europe(quotes []pokemontcgapi.Price) string {
	var rows []pokemontcgapi.Price
	for _, q := range raw(quotes) {
		if q.Source == "CARDMARKET" && q.Variant == "LOW" {
			rows = append(rows, q)
		}
	}
	if len(rows) == 0 {
		return noQuote
	}
	slices.SortStableFunc(rows, func(a, b pokemontcgapi.Price) int {
		return slices.Index(locales, locale(a)) - slices.Index(locales, locale(b))
	})
	lines := make([]string, 0, len(rows)+1)
	for _, q := range rows {
		code := locale(q)
		if code == "" {
			code = "??"
		}
		lines = append(lines, fmt.Sprintf("%s %.2f %s · %s · %s", strings.ToUpper(code), q.Amount, q.Currency, strings.ToLower(q.Basis), q.AsOf))
	}
	lines = append(lines, "Cardmarket, lowest asking price per print language")
	return strings.Join(lines, "\n")
}

func unitedStates(quotes []pokemontcgapi.Price) string {
	for _, q := range raw(quotes) {
		if q.Source == "TCGPLAYER" {
			return fmt.Sprintf("%.2f %s · %s · %s · %s", q.Amount, q.Currency, strings.ToLower(q.Basis), q.AsOf, q.Provenance)
		}
	}
	return noQuote
}

func footer(index *pokemontcgapi.PriceIndex) string {
	if index == nil {
		return "No index for this card"
	}
	return fmt.Sprintf("Index %.2f EUR · %s", index.EUR, index.AsOf)
}

func art(card *pokemontcgapi.Card) string {
	for _, i := range card.Images {
		if i.Size == "LARGE" {
			return i.URL
		}
	}
	if len(card.Images) > 0 {
		return card.Images[0].URL
	}
	return ""
}

File: typescript/format.ts
// The reply contract, shared by every bot in this repository: the Python and
// Go versions print byte-identical strings from the same fixtures, and the
// tests in each language assert it. Change one, change all three.

import type { Card, CardPrices, Price } from '@pokemontcgapi/sdk';

// One Cardmarket row per print language: an English copy and a French copy of
// the same card are two listings, not one price. The trial plan serves the
// English row only and says so in `meta.withheld`.
const LOCALES = ['en', 'de', 'fr', 'es', 'it', 'pt', 'ja', 'zh', 'ko'];

export const NO_QUOTE = 'no quote today';

export function title(card: Pick<Card, 'name' | 'set_name' | 'number'>): string {
  return `${card.name} · ${card.set_name} #${card.number}`;
}

// Raw copies only. A graded row is a different market, and on the API it is a
// different plan: a bigger bot gives slabs their own command.
export function raw(quotes: readonly Price[]): Price[] {
  return quotes.filter((q) => q.grading === null);
}

export function europe(quotes: readonly Price[]): string {
  const rows = raw(quotes)
    .filter((q) => q.source === 'CARDMARKET' && q.variant === 'LOW')
    .sort((a, b) => LOCALES.indexOf(a.locale ?? '') - LOCALES.indexOf(b.locale ?? ''));
  if (rows.length === 0) return NO_QUOTE;
  const lines = rows.map(
    (q) => `${(q.locale ?? '??').toUpperCase()} ${q.amount.toFixed(2)} ${q.currency} · ${q.basis.toLowerCase()} · ${q.as_of}`,
  );
  return `${lines.join('\n')}\nCardmarket, lowest asking price per print language`;
}

export function unitedStates(quotes: readonly Price[]): string {
  const q = raw(quotes).find((row) => row.source === 'TCGPLAYER');
  return q ? `${q.amount.toFixed(2)} ${q.currency} · ${q.basis.toLowerCase()} · ${q.as_of} · ${q.provenance}` : NO_QUOTE;
}

export function footer(index: CardPrices['index']): string {
  return index ? `Index ${index.eur.toFixed(2)} EUR · ${index.as_of}` : 'No index for this card';
}

export function art(card: Pick<Card, 'images'>): string | undefined {
  const image = card.images?.find((i) => i.size === 'LARGE') ?? card.images?.[0];
  return image?.url;
}
